### DISCUSSION TITLE
Analysis With Regular Expressions

### DISCUSSION THREAD

**Discussion by sgnl-labs** [pmd/pmd#5183] [Q&A]
Hi there folks. I have something on my plate right now that has me wondering what the community thinks. For the record, we are analyzing Apex and I've built quite a few custom Java rules that are in play right now.

To provide more flexible analysis options, our team would like to implement pattern matching via regular expressions. This gives our review/quality team the better automation and guards against human review error. 

The way I see it, there are a few ways I could implement this:
* A Java rule that has a couple of properties defining the pattern to be checked against, and another allowing for scope definition (class source vs. method source, that kind of thing.)
* If it's not expensive, have a rule that runs against the ASTApexFile source in question. This could get ugly of course, but it's the simplest option.
* Leverage a new XPath PMD function, like so:
```
//[element scope]/pmd-apex:bodyMatches([RegEx])
```

Thoughts?

**Comment by jsotuyod** [comment]
Hi! Thanks for bringing this up.

In all honesty, I feel this is not a great idea. The usage of the AST allows PMD to remove a lot of the grammar complexity from the rule definition (ie: `if`s that may or may not use braces, may be single-line or multi-line, etc.). My gut feeling is that this attempts to "reduce" the complexity of defining ad-hoc rules, as a quick and dirty way to detect a particular pattern; but it will fail to deal with the number of edge cases the language actually allows for.

Officially, XPath is the intended tool to achieve this. We get a nice balance of having a semantic way to describe the code structure, that allows to create new rules without needing to write code or package them in any way (even allow per-project rules to be defined in a local ruleset). If users are failing to perceive XPath as that "easy to use" tool, I believe this is great feedback, but we would have to better understand why.

Do you have any insights into why regexs would be preferred? what are the particular scenarios they are thinking about for this that can't be easily done with XPath?

**Reply by sgnl-labs** [reply]
Thanks for the quick response! Let me circle back with our team and I'll most certainly keep you in the loop. I think one of the big concerns is that RegEx is older than dirt and widely understood whereas XPath is less well known, not as well documented and the differences between v2/3 are staggering.

**Reply by jsotuyod** [reply]
looking forward to that answer!
In the meantime… have they looked into the [PMD Designer](https://docs.pmd-code.org/latest/pmd_userdocs_extending_designer_reference.html)? It's a visual tool specifically designed to build and test XPath rules.

**Reply by sgnl-labs** [reply]
Oh yes. We have made great use of the designer. If I told you how many rules our team has built out, you'd probably have a heart attack :)

**Reply by sgnl-labs** [reply]
So my assumption was correct in the grander sense I think; XPath is not an elegant query dialect to play around with. I've had version inconsistencies, and it does have some limitations regarding looking at elements outside of current scope. RegEx can do weird look-behinds, etc. But I get what you're saying. It's the AST first, and the implementation isn't as important as base structural issues

**Reply by sgnl-labs** [reply]
Also...I understand the idea that the code is analyzed as a structure. It's not a bunch of string juggling and comparison. The RegEx breaks that immersion, if you will.

**Reply by sgnl-labs** [reply]
I do think there might be a place for a RegEx XPath function though...for smaller inspections

**Reply by adangel** [reply]
Yes, we currently don't have the full text of node/file available in XPath. It's only available in Java-based rules, where you can call [getText()](https://docs.pmd-code.org/apidocs/pmd-core/7.4.0/net/sourceforge/pmd/lang/ast/TextAvailableNode.html#getText()) on basically any node. If you call it on the root node, you should get the complete underlying source text - on other nodes, you only get portions of it.
Maybe we removed that getter from XPath access because of memory concerns (the source text would appear at least partially multiple times), but not sure.

We use getText() in one rule: in PLSQL's [AvoidTabCharacter](https://docs.pmd-code.org/latest/pmd_rules_plsql_codestyle.html#avoidtabcharacter), where we iterate over each line.

IMHO, regex make sense, if the structure/syntax doesn't matter. In that sense, the mentioned rule will find a legit tab character inside a string literal, which is wrong (it should maybe only check leading/trailing whitespace). But with just the text, you can't decide that. We would need to provide some helper methods to go from a text position back into the AST (e.g. find node that contains a specific offset). This sounds like it wouldn't be so easy anymore in plain XPath and "just" executing a regex against the text.

Some rules might be implemented by just looking at the text - I think finding trojan characters (#3934) seems to be possible, as you just want to ban certain characters regardless where they occur. Or verifying a license header at the beginning of the file. But for verifying indentation in PLSQL [CodeFormat](https://docs.pmd-code.org/latest/pmd_rules_plsql_codestyle.html#codeformat) we don't use the text, rather we use the location information of the AST nodes - as indentation depends on structure.

I think, this request goes into the direction, that came up in the past: for Java rules, you need to package the classfile and put it onto the classpath, while for XPath rule, you just need the ruleset xml. It would be great to have something like "groovy based rules", so you would provide the implementation in the ruleset xml. Currently we have java-based rules, xpath-based rules. We could add regex based rules and groovy (or other simple scripting language) based rules. Refs #4321 and https://github.com/pmd/pmd/wiki/PMD-7.0.0-API#rule-api (search for "Rule Implementation Types").

**Reply by justinstroud-campmeeker** [reply]
Thanks for the responses. This is great and useful information.

I do like the idea of different dialects...it's funny you brought that up, because I was literally just ruminating on it a few minutes ago. What I was thinking was even a little more out there. 

What about JavaScript? :)

While it's everyone's favorite punching bag (myself included) it would be interesting (and in keeping with structural vs. lexical analysis) to have to option to use an OO-like approach to defining rules that gives you the flexibility of Java-based rules with the quick turnaround of XPath rules. This also opens up other possibilities, like TypeScript, if it's done right.

I'll check out the links you provided.

**Reply by justinstroud-campmeeker** [reply]
I'd go so far as saying "scriptable" rules. XPath is hard to put together when you're trying to access things that might be outside of your current scope. Accessing the AST via a script, though, is another story. While you may not get things like node streams out of the box, they could be built in. Where there's a will...

I don't have much recent experience embedding scripting into Java, but (even if it's just for fun) I feel like investigating. I know Nashorn was put to pasture, and even the OpenJDK version is in a state of neglect. That doesn't mean it won't work as a proof of concept though :)

**Reply by oowekyala** [reply]
Groovy is a scripting language and is very compatible with java. I think you will have a better time using that language (or clojure, or any jvm-first scripting language) than trying to integrate with Javascript. I think it would be a large burden to have to write and maintain JavaScript or typescript wrappers for our java APIs.

**Reply by sgnl-labs** [reply]
EDIT:

>  (or clojure, or any jvm-first scripting language)

Fortunately, Nashorn runs Javascript natively on the JVM. :)

I just integrated the OpenJDK version of Nashorn within PMD :) With full access to the AST node for the visit method that calls the source. Even made the JavaScript a property. 

No wrappers. Just the Maven dependency.

(sorry if there is confusion, @justinstroudbah and @sgnl-labs are the same person!)

CAVEAT (of course): Needs to be profiled and tested. This is a small POC that worked, and there is probably a better way to instance the engine and all that. But, the POC proved to my eyes that it can be integrated into PMD, it can run and consume a ASTUserClass instance, and it can give Java back a boolean result value. If needed, it can also return more complex information like values to be used in the rule message.

UPDATE:
So I executed this Javascript function from a PMD Java rule (please don't judge the nested loops, I was just tinkering):
```Javascript
                function hasViolation(node,forceResult){
                    var methodList = node.getMethods().toList();
                    for(var idx = 0; idx < methodList.size(); idx++){
                        print(methodList[idx].toString());
                        print(methodList[idx].getImage());
                        var childrenList = methodList[idx].children().toList();
                        for(var childIdx = 0; childIdx < childrenList.size(); childIdx++){
                            print('\t' + childrenList[childIdx].toString());
                            print('\t' + childrenList[childIdx].getImage());
                        }

                    }
                    print(node.getModifiers().toString());
                    return(forceResult);
                }
```
It pitched a little fit when I tried to use node streams, but that is to be expected. In a simple sense, it has access to the AST nodes and the underlying methods. I was able to use the return value in the Java rule to flag a violation.

Here is the Gist for the rule's visit method:
https://gist.github.com/justinstroudbah/17790f569c731d988ec62b238b93d7df


I'm kind of shocked how simple this all was.
